compiler/fir/checkers/checkers.jvm/src/org/jetbrains/kotlin/fir/analysis/jvm/checkers/expression/FirJavaGenericVarianceViolationTypeChecker.kt

```diff
@@ -0,0 +1,99 @@
+ * catches a common mistake allowed by flexible type.
+ *
+ * Consider a Java method accepting type `List<Object>`, which is mapped to `MutableList<Any?>..List<(out) Any?>`. If one passes a mutable
+ * list of some more concrete type than `Any` (ArrayList<String>, LinkedList<Int>, etc.) to this Java method, then any writes to this
+ * mutable collection could cause `ClassCastException`s if the same mutable list is read elsewhere. It's the purpose of this checker to
+ * reject such code.
+ *
+ * On the other hand, if one passes an immutable list of some more concrete type, then any writes to it on the Java side would cause
+ * `UnsupportedOperationException`, which is expected and the price we pay in order to make immutable collection easier to use. This checker
+ * doesn't do anything to prevent this from happening.
+ */
+object FirJavaGenericVarianceViolationTypeChecker : FirFunctionCallChecker() {
+    private val javaOrigin = setOf(FirDeclarationOrigin.Java, FirDeclarationOrigin.Enhancement)
+
+    override fun check(expression: FirFunctionCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        val calleeFunction = expression.calleeReference.toResolvedCallableSymbol()?.fir as? FirFunction<*> ?: return
+        if (calleeFunction.origin !in javaOrigin) {
+            return
+        }
+        val argList = expression.argumentList as? FirArgumentListWithResolutionInfo
+        val typeParameterSubstitutor = argList?.typeParameterSubstitutor ?: return
+        for ((arg, param) in argList.mapping) {
+            val expectedType = typeParameterSubstitutor.substituteOrSelf(param.returnTypeRef.coneType)
+
+            // optimization: if no arguments or flexibility, everything is OK
+            if (expectedType !is ConeFlexibleType || expectedType.typeArguments.isEmpty()) continue
+
+            // Anything is acceptable for raw types
+            if (expectedType is ConeRawType) continue
+
+            val lowerBound = expectedType.lowerBound
+            val upperBound = expectedType.upperBound
+            val typeCtx = context.session.inferenceComponents.ctx
+            val lowerConstructor = lowerBound.typeConstructor(typeCtx)
+            val upperConstructor = upperBound.typeConstructor(typeCtx)
+
+            // Use site variance projection is always the same for flexible types. So there is no need to check if declaration site is the
+            // same.
+            if (lowerConstructor == upperConstructor) return
+
+            val argType = arg.typeRef.coneType
+
+            // If the base class of the argument type is not equal or a sub class of the lower bound, then we simply allow it so that
+            // Kotlin immutable collections can be used in place of Java collection types.
+            if (!typeCtx.isTypeConstructorEqualOrSubClassOf(argType, lowerBound)) return
+
+            if (!AbstractTypeChecker.isSubtypeOf(typeCtx, argType, lowerBound)) {
```

Comment:
`JavaGenericVarianceViolationTypeChecker` has a bit more complicated logic just because of the cases above